### PR TITLE
Add lesson on formatting and linting

### PR DESCRIPTION
**Pull request by keithchev**
This PR adds a lesson on formatting, style guides, and linting. This is intended for the AUG lesson scheduled for 2024-02-06. 

I mostly wrote this lesson from scratch myself, with some material and suggestions from copilot and GPT4. I also borrowed some examples from the carpentries lessons on [style guides](https://carpentries-incubator.github.io/python-intermediate-development-earth-sciences/15-coding-conventions/index.html) and [linting](https://carpentries-incubator.github.io/python-intermediate-development-earth-sciences/16-verifying-code-style-linters/index.html).

I tried to make the lesson somewhat interactive, though it could probably be made more so. I'm also worried there's too much material to cover in an hour and that the text turned out rather too dense and verbose. Please let me know what you think!

### CODE REVIEW

File: docs/arcadia-users-group/20240206-intro-to-formatting-and-linting/lesson.md [hunk 1]
```diff
@@ -0,0 +1,425 @@
...
+    full_name = first_name + last_nam
+    print(f'Your full name is {full_name}!')
+```
+Note that there is a typo in the function: the variable `last_name` is misspelled as `last_nam`. Unfortunately, because Python is interpreted, this mistake will go unnoticed until the code is run and the `show_full_name` function is called. At this point, an error will occur (and the Python process will crash) because the variable `last_nam` is not defined. A linter, by analyzing the source code, can detect this kind of error before the code is ever run (much less shared with others or released) and flag it for the developer to fix.
+
+In Python, one popular linting tool is called `ruff`. Let's try it out. First, install `ruff` in our conda environment:
+```bash
+mamba install -y ruff
+```
+Now, copy the function above into `main.py` and run `ruff` from the command line:
```
Comment (taylorreiter) [inline docs/arcadia-users-group/20240206-intro-to-formatting-and-linting/lesson.md:377]: so you'll make sure learners are on the lesson and scrolling along with you, or post this into the slack chat so that everyone can have a copy?

File: docs/arcadia-users-group/20240206-intro-to-formatting-and-linting/lesson.md [hunk 2]
```diff
@@ -0,0 +1,425 @@
...
+
+In general, choosing the appropriate set of lint rules to enforce is a context- and project-dependent decision. As codebases grow and either become more complex or are developed by multiple people, it often makes sense to enforce increasingly strict and extensive linting rules. For our purposes at Arcadia, using the default rules that `ruff` enforces (i.e., using `ruff check` alone) is a great place to start; this will catch many common typos and bugs that will cause errors at runtime (like undefined and unused variables). 
+
+
+### When to format and when to lint
+The short answer is early and often: because formatters and linters are fast and easy to run, it is best to run them frequently and as early as possible in the development process, when errors are easy to fix. Most IDEs can be configured to run formatters automatically each time a file is saved, and linters are often integrate with IDEs as well (this is how VS Code, for example, displays squiggly red lines under undefined variables).
```
Comment (taylorreiter) [inline docs/arcadia-users-group/20240206-intro-to-formatting-and-linting/lesson.md:419]:
```suggestion
The short answer is early and often: because formatters and linters are fast and easy to run, it is best to run them frequently and as early as possible in the development process, when errors are easy to fix. Most IDEs can be configured to run formatters automatically each time a file is saved, and linters are often integrated with IDEs as well (this is how VS Code, for example, displays squiggly red lines under undefined variables).
```

File: docs/arcadia-users-group/20240206-intro-to-formatting-and-linting/lesson.md [hunk 3]
```diff
@@ -0,0 +1,425 @@
...
+
+In addition, formatting and linting are usually run automatically as part of a continuous integration (CI) pipeline. Briefly, this means that, for example, whenever a PR is opened on GitHub, the same formatting and linting tools that a developer would run locally are run remotely on the code in the PR. This ensures that the code on the `main` branch in the GitHub repo--which is the "final" version of the code that will ultimately be shared with or deployed to users--is properly formatted and passes the project's linting rules, whether or not individual developers took the time to run these tools locally.
```
Comment (taylorreiter) [inline docs/arcadia-users-group/20240206-intro-to-formatting-and-linting/lesson.md:421]:
```suggestion
In addition, formatting and linting are usually run automatically as part of a [continuous integration (CI) pipeline](https://training.arcadiascience.com/arcadia-users-group/20231104-testing-concepts/lesson/#continuous-integration). Briefly, this means that, for example, whenever a PR is opened on GitHub, the same formatting and linting tools that a developer would run locally are run remotely on the code in the PR. This ensures that the code on the `main` branch in the GitHub repo--which is the "final" version of the code that will ultimately be shared with or deployed to users--is properly formatted and passes the project's linting rules, whether or not individual developers took the time to run these tools locally. 
```

File: docs/arcadia-users-group/20240206-intro-to-formatting-and-linting/lesson.md [hunk 4]
```diff
@@ -0,0 +1,425 @@
...
+    # find the minimum value and clamp it to 0.
+    min_value = max(min(values), 0)
+
+    # get all the .txt files in the input directory.
+    filepaths = input_dirpath.glob('*.txt')
+
+    # create the directory if it doesn't already exist.
+    if not os.path.exists(dirpath):
+        os.mkdir(dirpath)
+    ```
```
Comment (taylorreiter) [inline docs/arcadia-users-group/20240206-intro-to-formatting-and-linting/lesson.md:310]: this block doesn't format correctly when i viewed the page with `mkdocs serve`. I think it needs to be not indented? I think it will also mess up your numbering scheme potentially, but im not sure. 

I tried to make the suggested change but the triple backticks messed it up and wouldn't let me :(

**Review by taylorreiter — APPROVED** [review]
This is really wonderful, thank you for putting this together Keith! I learned a lot reading it, and i felt like all of the examples are approachable.

I corrected a few typos, suggested some places for links (or added them myself), and made a few comments about when i think things will be hard to teach or material that can be only in the lesson. Feel free to take/reject whichever suggestions feel right to you.

File: docs/arcadia-users-group/20240206-intro-to-formatting-and-linting/lesson.md [hunk 5]
```diff
@@ -0,0 +1,425 @@
...
+    root_2 = (-linear_coeff - discriminant_square_root) / (2 * quadratic_coeff)
+
+    return root_1, root_2
+```
+Notice how this code is easier to read and easier to understand (if still somewhat opaque). Notice that there are several kinds of changes in this version. For one, spaces are used around the operators, which generally makes code easier to read; this is a kind of *formatting convention*. In addition, the variable names are much longer and are more descriptive, which make it easier for reader to infer the meaning of the code; this is a kind of *style convention*. 
+
+In this lesson, we'll discuss both formatting and style conventions, as well as tools to automatically format code and check that is complies with our style conventions.
+
+## Lesson setup
+We'll create a simple Python project to demonstrate how to use the tools we'll be discussing.
```
[the remark this answers is not in the document]
Comment (keithchev) [reply docs/arcadia-users-group/20240206-intro-to-formatting-and-linting/lesson.md:39]: Thank you! I try to be super mindful of this but obviously I sometimes fall short 😬

File: docs/arcadia-users-group/20240206-intro-to-formatting-and-linting/lesson.md [hunk 6]
```diff
@@ -0,0 +1,425 @@
...
+```python
+def calculate_total_income(gross_wages, taxable_interest, total_dividends, qualified_dividends, other_income):
+    total_income = gross_wages + taxable_interest + (total_dividends - qualified_dividends) + other_income
+    return total_income
+```
+Now, from the command line, run:
+```bash
+black .
+```
+This command tells `black` to reformat all the Python files in the current directory (the lone dot in the command is the relative path to the current directory). If you look at `main.py` now, you'll see that the code has been reformatted according to `black`'s rules.
```
[the remark this answers is not in the document]
Comment (keithchev) [reply docs/arcadia-users-group/20240206-intro-to-formatting-and-linting/lesson.md:127]: Well the dot itself is not looking for anything - it just represents the current directory, which black then searches for python files. Do you mean what files does black look for? afaik it searches for all of files with a `.py` extension in the directory (or its subdirectories). This is what I meant by "all of the python files in the directory" but I can add this to be more explicit.

File: docs/arcadia-users-group/20240206-intro-to-formatting-and-linting/lesson.md [hunk 7]
```diff
@@ -0,0 +1,425 @@
...
+
+Now, try simplifying the calculation of `total_income` by removing the dividends from the sum, then run `black` again. You should see that, because the `total_income` line is shorter, it can fit on one line, so `black` automatically converts it back to a single line.
```
[the remark this answers is not in the document]
Comment (keithchev) [reply docs/arcadia-users-group/20240206-intro-to-formatting-and-linting/lesson.md:129]: I was planning to have everyone try this themselves and also demo it live myself. I was hoping that by not showing the output ahead of time in the text, the result would be a surprise and that this would implicitly emphasize that `black` is totally automated and also that there's no need to know or care exactly how it'll format something in advance. Does that make sense?

File: docs/arcadia-users-group/20240206-intro-to-formatting-and-linting/lesson.md [hunk 8]
```diff
@@ -0,0 +1,425 @@
...
+
+### PEP8 and Python style guides
+Many language have an official style guide. For Python, the official style guide is called [PEP8](https://peps.python.org/pep-0008/#introduction). In addition, Google's [Python style guide](https://google.github.io/styleguide/pyguide.html) is an important extension of PEP8 that defines additional standards, especially around comments and documentation. Most major Python projects follow at least PEP8, and we should strive to follow it at Arcadia as well. This is true no matter the scale of the project; it is a good practice to follow PEP8 even for small scripts and one-off analyses. As projects grow, it is usually wise to start imposing additional style conventions from Google' style guide on top of PEP8.
```
[the remark this answers is not in the document]
Comment (keithchev) [reply docs/arcadia-users-group/20240206-intro-to-formatting-and-linting/lesson.md:138]: Ah good point! I'll add that. I think it'll be much more effective to provide a specific motivation, rather than a vague "do this or else" prescription.

File: docs/arcadia-users-group/20240206-intro-to-formatting-and-linting/lesson.md [hunk 9]
```diff
@@ -0,0 +1,425 @@
...
+#### Aside about single-letter variable names
+Although PEP8 and other style guides do not explicitly forbid the use of single-letter variable names, they are strongly discouraged in most contexts. This is because single-letter variable names are, by definition, not descriptive. They can also be quite literally ambiguous; it is often hard to visually detect the difference between `i`, `j`, `l`, and `1`, for example. Finally, a subtler reason to avoid single-character variable names is that they can make refactoring harder, as it may be more cumbersome to search for and rename single-character variable names than more descriptive ones.
+
+The only exception to this general prohibition should be for variables defined in a very local or narrow scope such as a short `for` loop or in a lambda function. Here are two examples where single-letter variable names is used within a single line (which is as narrow a scope as possible):
+```python
+# single-letter variable names in a list comprehension
+filepaths = ['data_1.txt', 'data_2.txt', 'data_3.csv']
+txt_filepaths = [f for f in filepaths if f.endswith('.txt')]
+
+# single-letter variable names in an anonymous lambda function
```
[the remark this answers is not in the document]
Comment (keithchev) [reply docs/arcadia-users-group/20240206-intro-to-formatting-and-linting/lesson.md:223]: yeah, I think a lambda function is by definition anonymous, so I'll just remove the "anonymous" part.

File: docs/arcadia-users-group/20240206-intro-to-formatting-and-linting/lesson.md [hunk 10]
```diff
@@ -0,0 +1,425 @@
...
+num_timepoints, num_rows, num_cols = image.shape
+for time_ind in range(num_timepoints):
+    for row_ind in range(num_rows):
+        for col_ind in range(num_cols):
+            intensity = image[time_ind, row_ind, col_ind]
+            print(
+                f'The pixel intensity at timepoint {time_ind} '
+                f'and position ({row_ind}, {col_ind}) is {intensity}'
+            )
+```
```
[the remark this answers is not in the document]
Comment (keithchev) [reply docs/arcadia-users-group/20240206-intro-to-formatting-and-linting/lesson.md:266]: Good idea! That'll also make it easier to skip if we're running short on time.

File: docs/arcadia-users-group/20240206-intro-to-formatting-and-linting/lesson.md (hunk 4 above)
Comment (keithchev) [reply docs/arcadia-users-group/20240206-intro-to-formatting-and-linting/lesson.md:315]: oof that's annoying. I'll just rewrite this to eliminate the need to use numbering (and therefore the indents).

File: docs/arcadia-users-group/20240206-intro-to-formatting-and-linting/lesson.md (hunk 1 above)
Comment (keithchev) [reply docs/arcadia-users-group/20240206-intro-to-formatting-and-linting/lesson.md:381]: Good question! I can have it ready to copy-paste into slack and the zoom chat.

File: docs/arcadia-users-group/20240206-intro-to-formatting-and-linting/lesson.md [hunk 11]
```diff
@@ -0,0 +1,429 @@
...
+The short answer is early and often: because formatters and linters are fast and easy to run, it is best to run them frequently and as early as possible in the development process, when errors are easy to fix. Most IDEs can be configured to run formatters automatically each time a file is saved, and linters are often integrated with IDEs as well (this is how VS Code, for example, displays squiggly red lines under undefined variables).
+
+In addition, formatting and linting are usually run automatically as part of a [continuous integration (CI) pipeline](https://training.arcadiascience.com/arcadia-users-group/20231104-testing-concepts/lesson/#continuous-integration). Briefly, this means that, for example, whenever a PR is opened on GitHub, the same formatting and linting tools that a developer would run locally are run remotely on the code in the PR. This ensures that the code on the `main` branch in the GitHub repo--which is the "final" version of the code that will ultimately be shared with or deployed to users--is properly formatted and passes the project's linting rules, whether or not individual developers took the time to run these tools locally.
```
Comment (mezarque) [inline docs/arcadia-users-group/20240206-intro-to-formatting-and-linting/lesson.md:425]: Might be good to tease here the upcoming AUG on GitHub templates!

I think it might also make people feel better if we communicate that the responsibility of formatting/ linting doesn't fall on the individual alone, but that we're working on ways (such as through the GitHub templates) to make this class of tasks as painless as possible.

**Review by mezarque — APPROVED** [review]
This is great, thanks for putting this all together! Just a few small typos.

The one issue I foresee will happen is that a lot of the details here are relevant to Python (e.g. the exact formatters, linters, type hinting, etc.). Having started to explore some of this stuff in R, which is pretty commonly used by Arcadians, it's a much messier and less user-friendly world.

Even if not in this lesson, we should probably put together some guidance for how to get some of these practices up and running for R users. Even just setting up VSCode to get syntax highlighting to work for R has been a headache for me (and Austin).

File: docs/arcadia-users-group/20240206-intro-to-formatting-and-linting/lesson.md [hunk 12]
```diff
@@ -0,0 +1,429 @@
...
+```
+
+#### Things comments should *not* be used for
+There are two major ways in which comments are commonly misused. Note that avoiding these misuses is not specific to Python but is a general best practice that applies to all programming languages. The first is that, as we alluded to above, comments should not be redundant with the code. They should not simply restate what the code is doing and they should not include information that is apparent or readily inferred from the code itself. Here are some examples of such redundant comments:
+```python
+# find the minimum value and clamp it to 0.
+min_value = max(min(values), 0)
+
+# get all the .txt files in the input directory.
+filepaths = input_dirpath.glob('*.txt')
```
[the remark this answers is not in the document]
Comment (keithchev) [reply docs/arcadia-users-group/20240206-intro-to-formatting-and-linting/lesson.md:308]: Yeah, good point. I'll add a little disclaimer here that the definition of "redundant" should depend to some extent on the audience. I would say that it's still *generally* best to err on the side of avoiding redundancy, at least when commenting on the behavior of third-party modules (or properly-documented first-party modules), as IDEs are pretty good at unobtrusively surfacing their documentation (e.g. VS Code shows tooltips on hover with the docstring summary).

File: docs/arcadia-users-group/20240206-intro-to-formatting-and-linting/lesson.md (hunk 11 above)
Comment (keithchev) [reply docs/arcadia-users-group/20240206-intro-to-formatting-and-linting/lesson.md:425]: Yes, totally.

**Comment by keithchev** [comment]
> Even if not in this lesson, we should probably put together some guidance for how to get some of these practices up and running for R users. Even just setting up VSCode to get syntax highlighting to work for R has been a headache for me (and Austin).

Agreed! I know nothing about the R ecosystem but can imagine this stuff is not as straightforward as it is in Python. But it would be great to have *something*. Even simple checks like catching undefined and unused variables go a long way...

File: docs/arcadia-users-group/20240206-intro-to-formatting-and-linting/lesson.md [hunk 13]
```diff
@@ -0,0 +1,433 @@
...
+
+This is true no matter the scale of the project; it is a good practice to follow PEP8 even for small scripts and one-off analyses, as it is often hard to anticipate the lifetime of a project in advance. It is therefore prudent to assume that all of the code we write will always be read by others (inside, if not also outside, of Arcadia). As projects grow and it becomes more likely that we will need to maintain them for a long period of time, we can then readily impose additional style conventions from Google's style guide on top of PEP8.
+
+Many of the conventions defined in PEP8 concern formatting and are automatically enforced by using `black` to format our code. However, PEP8 also includes style conventions that we must follow manually. Here, we'll discuss two important categories of style conventions: naming conventions and documentation conventions.
```
Comment (giladmishne) [inline docs/arcadia-users-group/20240206-intro-to-formatting-and-linting/lesson.md:144]: would it be useful to mention exceptions/pragmas to both formatting and linting, esp since scientific code often has a few of these? items that come to mind are forcing multi-line lists with trailing commas on the last item, allowing the occasional uppercase in a variable e.g. for chemical compound names, etc.

File: docs/arcadia-users-group/20240206-intro-to-formatting-and-linting/lesson.md [hunk 14]
```diff
@@ -0,0 +1,433 @@
...
+```
+Now, from the command line, run:
+```bash
+black .
+```
+This command tells `black` to reformat all the Python files in the current directory (the lone dot in the command is the relative path to the current directory). If you look at `main.py` now, you'll see that the code has been reformatted according to `black`'s rules.
+
+Now, try simplifying the calculation of `total_income` by removing the dividends from the sum, then run `black` again. You should see that, because the `total_income` line is shorter, it can fit on one line, so `black` automatically converts it back to a single line.
+
+There is a succinct overview of `black`'s formatting rules in [its documentation](https://black.readthedocs.io/en/stable/the_black_code_style/current_style.html). While it's good to be familiar with these rules, there is no need to memorize them. Indeed, the purpose of using a formatting tool like `black` is to eliminate the need to manually write your code according to any particular formatting conventions or even to think about formatting at all.
```
Comment (giladmishne) [inline docs/arcadia-users-group/20240206-intro-to-formatting-and-linting/lesson.md:133]: since many people write primarily in notebooks, mention `nb_black`? or if they do it in VScode I guess it's built in?

**Review by giladmishne — APPROVED** [review]
This is a great lesson. Clear examples with the occasional nod to typical code at Arcadia, and covers all the important aspects.

File: docs/arcadia-users-group/20240206-intro-to-formatting-and-linting/lesson.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (taylorreiter) [reply docs/arcadia-users-group/20240206-intro-to-formatting-and-linting/lesson.md:129]: sorry yes i meant what files black looks for, and yes i meant whether it searches for only `.py` extensions or other extensions as well, and whether it checks subdirectories. I think adding those two bits would be helpful context for people to know!

File: docs/arcadia-users-group/20240206-intro-to-formatting-and-linting/lesson.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (taylorreiter) [reply docs/arcadia-users-group/20240206-intro-to-formatting-and-linting/lesson.md:131]: yes! that sounds great.

File: docs/arcadia-users-group/20240206-intro-to-formatting-and-linting/lesson.md [hunk 15]
```diff
@@ -0,0 +1,433 @@
...
+    if n <= 0:
+        return 0
+    elif n == 1:
+        return 1
+    else:
+        return fibonacci(n - 1) + fibonacci(n - 2)
+```
+
+#### Things comments should *not* be used for
+There are two major ways in which comments are commonly misused. Note that avoiding these misuses is not specific to Python but is a general best practice that applies to all programming languages. The first is that, as we alluded to above, comments should not be redundant with the code. They should not simply restate what the code is doing and they should not include information that is apparent or readily inferred from the code itself. Here are some examples of such redundant comments:
```
[the remark this answers is not in the document]
Comment (taylorreiter) [reply docs/arcadia-users-group/20240206-intro-to-formatting-and-linting/lesson.md:302]: ooh this is a really good point.

File: docs/arcadia-users-group/20240206-intro-to-formatting-and-linting/lesson.md [hunk 16]
```diff
@@ -0,0 +1,433 @@
...
+    discriminant_square_root = np.sqrt(
+        linear_coeff ** 2 - 4 * quadratic_coeff * constant_coeff
+    )
+
+    root_1 = (-linear_coeff + discriminant_square_root) / (2 * quadratic_coeff)
+    root_2 = (-linear_coeff - discriminant_square_root) / (2 * quadratic_coeff)
+
+    return root_1, root_2
+```
+Notice how this code is easier to read and easier to understand (if still somewhat opaque). Notice that there are several kinds of changes in this version. For one, spaces are used around the operators, which generally makes code easier to read; this is a kind of *formatting convention*. In addition, the variable names are much longer and are more descriptive, which make it easier for reader to infer the meaning of the code; this is a kind of *style convention*.
```
Comment (mertcelebi) [inline docs/arcadia-users-group/20240206-intro-to-formatting-and-linting/lesson.md:36]: Sometimes people optimize for line counts ("oh I did this thing in a one-liner" type of attitude). And maybe calling that out specifically and saying in cases like this being more understandable is worth the cost of being less concise etc. 

Some languages are more-prone to this type of stuff. Like Ruby has a lot of syntactic sugar that makes it easy to write cute one-liners that are hard to parse

File: docs/arcadia-users-group/20240206-intro-to-formatting-and-linting/lesson.md [hunk 17]
```diff
@@ -0,0 +1,433 @@
...
+
+#### Comments
+Comments are human-readable lines of text that can appear anywhere in the source code. In Python, they are denoted by the `#` character. Although they are ignored by the Python interpreter and might seem like an area where "anything goes", it is important to adhere to strict standards of grammar and style when writing comments, since they are often the only way to explain the purpose of a particular line or block of code.
+
+In particular, comments should be complete sentences with proper punctuation. They should be written in full English sentences and should be grammatically correct. Importantly, they should end with periods; this is not only grammatically correct but is also the only way to indicate to the reader that the comment is complete and was not accidentally truncated (or never completely written in the first place). 
+
+Most importantly, comments should generally be used to explain *why* a particular line or block of code is doing what it is doing, not *what* it is doing or *how* it is doing it (usually, this should be apparent from the code itself).
```
Comment (mertcelebi) [inline docs/arcadia-users-group/20240206-intro-to-formatting-and-linting/lesson.md:277]: I'd probably expand this a little bit and say comments are good for drawing attention, so while the "why" is generally important, in cases where you're doing something unintuitive/weird/magical, it could be worth calling out the what as well as the why.

File: docs/arcadia-users-group/20240206-intro-to-formatting-and-linting/lesson.md [hunk 18]
```diff
@@ -0,0 +1,433 @@
...
+main.py:1:20: ANN001 Missing type annotation for function argument `first_name`
+main.py:1:44: ANN001 Missing type annotation for function argument `last_name`
+main.py:3:5: T201 `print` found
+Found 8 errors.
+```
+Notice that, among other issues, `ruff` has now noticed that our function (and indeed the `main.py` module itself) does not have a docstring. As an aside, the `warning` lines at the beginning of the output indicate a problem with our use of `ruff`: we have enabled so many lint rules that some of the rules are actually in conflict with one another, so `ruff` has to choose which ones to enforce. 
+
+In general, choosing the appropriate set of lint rules to enforce is a context- and project-dependent decision. As codebases grow and either become more complex or are developed by multiple people, it often makes sense to enforce increasingly strict and extensive linting rules. For our purposes at Arcadia, using the default rules that `ruff` enforces (i.e., using `ruff check` alone) is a great place to start; this will catch many common typos and bugs that will cause errors at runtime (like undefined and unused variables).
```
Comment (mertcelebi) [inline docs/arcadia-users-group/20240206-intro-to-formatting-and-linting/lesson.md:420]: I'd maybe link to our template repos for Python + Snakemake and mention they come all configured to work with ruff etc.

File: docs/arcadia-users-group/20240206-intro-to-formatting-and-linting/lesson.md [hunk 19]
```diff
@@ -0,0 +1,433 @@
...
+
+### Code formatters enforce formatting conventions
+To both eliminate these kinds of formatting ambiguities and to automate the otherwise-tedious process of formatting code by hand, tools to automatically format code have been developed. These tools take files of source code as input and output the same code, but reformatted as necessary according to a pre-specified set of formatting rules.
+
+In Python, the most popular code formatting tool is called `black`. This tool both automatically formats code and imposes its own opinionated set of formatting rules (hence its tagline: "any color you want"). This eliminates the need for developers to make (or argue about) these decisions themselves. Although `black`'s formatting rules can take some getting used to, the benefit of eliminating almost all formatting-related decisions is usually worth the effort of adjusting to its opinions. 
+
+
+### Setting up `black`
```
Comment (mertcelebi) [inline docs/arcadia-users-group/20240206-intro-to-formatting-and-linting/lesson.md:113]: Since we use ruff for linting and ruff is meant to be a drop in replacement for black, what is the reason not to cover this section with ruff? Educational? If so, I'd call out the fact that ruff can be used instead. 

And in the case of Jupyter notebooks, Ruff has native support so you wouldn't need nb-black etc.

**Review by mertcelebi — COMMENTED** [review]
Leaving a couple comments. Thanks for putting it together, looks very detailed.

File: docs/arcadia-users-group/20240206-intro-to-formatting-and-linting/lesson.md (hunk 15 above)
Comment (mertcelebi) [inline docs/arcadia-users-group/20240206-intro-to-formatting-and-linting/lesson.md:302]: Could be worth calling out here that this is the opposite of what ChatGPT does, which makes ChatGPT code very annoying to work with.

File: docs/arcadia-users-group/20240206-intro-to-formatting-and-linting/lesson.md (hunk 19 above)
Comment (mezarque) [reply docs/arcadia-users-group/20240206-intro-to-formatting-and-linting/lesson.md:113]: I was also a little confused by this; is it b/c we want an example of a formatter that is separate from a linter?

File: docs/arcadia-users-group/20240206-intro-to-formatting-and-linting/lesson.md [hunk 20]
```diff
@@ -0,0 +1,433 @@
...
+
+# good
+user_input = input('Enter your name: ')
+```
+
+
+#### Aside about single-letter variable names
+Although PEP8 and other style guides do not explicitly forbid the use of single-letter variable names, they are strongly discouraged in most contexts. This is because single-letter variable names are, by definition, not descriptive. They can also be quite literally ambiguous; it is often hard to visually detect the difference between `i`, `j`, `l`, and `1`, for example. Finally, a subtler reason to avoid single-character variable names is that they can make refactoring harder, as it may be more cumbersome to search for and rename single-character variable names than more descriptive ones.
+
+The only exception to this general prohibition should be for variables defined in a very local or narrow scope such as a short `for` loop or in a lambda function. Here are two examples where single-letter variable names is used within a single line (which is as narrow a scope as possible):
```
[the remark this answers is not in the document]
Comment (keithchev) [reply docs/arcadia-users-group/20240206-intro-to-formatting-and-linting/lesson.md:217]: Added!

File: docs/arcadia-users-group/20240206-intro-to-formatting-and-linting/lesson.md (hunk 15 above)
[the remark this answers is not in the document]
Comment (keithchev) [reply docs/arcadia-users-group/20240206-intro-to-formatting-and-linting/lesson.md:302]: Added!

File: docs/arcadia-users-group/20240206-intro-to-formatting-and-linting/lesson.md (hunk 14 above)
Comment (keithchev) [reply docs/arcadia-users-group/20240206-intro-to-formatting-and-linting/lesson.md:133]: I think it's builtin, and hopefully people are/will be writing notebooks in a repo that was started from our python-project template (which uses ruff, which has builtin support for notebooks)

File: docs/arcadia-users-group/20240206-intro-to-formatting-and-linting/lesson.md (hunk 19 above)
Comment (keithchev) [reply docs/arcadia-users-group/20240206-intro-to-formatting-and-linting/lesson.md:113]: I used `black` in the lesson because I thought it would be clearer and I wanted to give credit to `black` for its role in standardizing formatting in python. The fact that `ruff format` reimplements black and is faster feels like an implementation detail. But I will add an explicit mention of this at the end when we mention the template repos.

Comment (mertcelebi) [reply docs/arcadia-users-group/20240206-intro-to-formatting-and-linting/lesson.md:113]: Yeah, but we don't people to get to sense that they need both. So good to clarify

File: docs/arcadia-users-group/20240206-intro-to-formatting-and-linting/lesson.md (hunk 16 above)
Comment (keithchev) [reply docs/arcadia-users-group/20240206-intro-to-formatting-and-linting/lesson.md]: Added!

File: docs/arcadia-users-group/20240206-intro-to-formatting-and-linting/lesson.md (hunk 13 above)
Comment (keithchev) [reply docs/arcadia-users-group/20240206-intro-to-formatting-and-linting/lesson.md:151]: I think I'll leave this for people to ask about; I feel like there's already enough detail/density in the lesson.
